File: LeetCode/Problems-Python/41 First Missing Positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        
        i = 1
        while i <= len(nums):
            x = nums[i-1]
            if 1 <= x <= len(nums) and nums[x-1] != x:
                nums[i-1], nums[x-1] = nums[x-1], x
            else:
                i += 1
            
        for i in range(len(nums)):
            if nums[i]!=i+1:
                return i+1
        return len(nums) + 1
        
        
        
        
#         smallest = 1
#         while smallest in nums:
#             smallest+=1
            
                
        
#         numSet = set(nums)
        
#         while smallest in numSet:
#             smallest+=1
            
        # return smallest

File: LeetCode/Problems-CPP/41 First Missing Positive.cpp
// Time Taken(%):8ms
// Memory usage(%):10MB
// Your Name(optional):Aditya Upadhyay
class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        // int max_num = 0 ;
        // for(int i=0;i<nums.size();i++){
        //     if(nums[i]>0) max_num = max(max_num,nums[i]);
        // }
        // if(nums.size()==0) return 1;
        unordered_map<int,int> mp;
        for(int i=0;i<nums.size();i++){
            if(nums[i]>0) mp[nums[i]]=1;
        }
        for(int i=1;i<=2*nums.size();i++){
            if(mp.find(i)==mp.end()) return i;
        }
        return 1;
        // for(int i=0;i<nums.size();i++){
        //     if(nums[i]>0) arr[nums[i]]=true;
        // }
        // for(int i=1;i<max_num+1;i++){
        //     if(arr[i]==false) return i;
        // }
        // return max_num+1;        
    }
};